README.md
[![Build Status](https://travis-ci.org/platanus/lightning-prometheus-exporter.svg?branch=master)](https://travis-ci.org/platanus/lightning-prometheus-exporter) [![](https://images.microbadger.com/badges/version/platanus/lightning-prometheus-exporter.svg)](https:/hub.docker.com/r/platanus/lightning-prometheus-exporter) [![Go Report Card](https://goreportcard.com/badge/github.com/platanus/lightning-prometheus-exporter)](https://goreportcard.com/report/github.com/platanus/lightning-prometheus-exporter)

# lightning Prometheus Exporter

lightning Prometheus exporter makes it possible to monitor lightning node using Prometheus.

## Overview

Lightning Prometheus exporter fetches the metrics the rpc api of the node, converts the metrics into appropriate Prometheus metrics types and finally exposes them via an HTTP server to be collected by [Prometheus](https://prometheus.io/).

## Getting Started

In this section, we show how to quickly run lightning Prometheus Exporter for lightning.

### Prerequisites

We assume that you have already installed Prometheus and a LND Lightning Node. Additionally, you need to configure Prometheus to scrape metrics from the server with the exporter. Note that the default scrape port of the exporter is `9113` and the default metrics path -- `/metrics`.

## Usage

### Command-line Arguments

```
Usage of ./lightning-prometheus-exporter:
  -namepsace string
        The namespace or prefix to use in the exported metrics. The default value can be overwritten by NAMESPACE environment variable (default: lnd)
  -web.telemetry-path string
        A path under which to expose metrics. The default value can be overwritten by TELEMETRY_PATH environment variable. (default "/metrics")
  -web.listen-address string
        An address to listen on for web interface and telemetry. The default value can be overwritten by LISTEN_ADDRESS environment variable. (default ":9113")
  -rpc.host string
        Lightning node RPC host. The default value can be overwritten by RPC_HOST environment variable (default: localhost)
  -rpc.Port int
        Lightning node RPC port. The default value can be overwritten by RPC_PORT environment variable (default: 10009)
  -lnd.tls-cert-path string
        The path to the tls certificate. The default value can be overwritten by TLS_CERT_PATH environment variable (default: "/root/.lnd")
  -lnd.macaroon-path
        The path to the read only macaroon. The default value can be overwritten by MACAROON_PATH environment variable
  -go-metrics bool
        Enable process and go metrics from go client library. The default value can be overwritten by GO_METRICS environmental variable.
```

### Exported Metrics

* Connect to the `/metrics` page of the running exporter to see the complete list of metrics along with their descriptions.

### Troubleshooting

The exporter logs errors to the standard output. When using Docker, if the exporter doesn’t work as expected, check its logs using [docker logs](https://docs.docker.com/engine/reference/commandline/logs/) command.

## Releases

For each release, we publish the corresponding Docker image at `platanus/lightning-prometheus-exporter` [DockerHub repo](https://hub.docker.com/r/platanus/lightning-prometheus-exporter/) and the binaries on the GitHub [releases page](https://github.com/platanus/lightning-prometheus-exporter/releases).

## Building the Exporter

You can build the exporter using the provided Makefile. Before building the exporter, make sure the following software is installed on your machine:
* make
* git
* Docker for building the container image
* Go for building the binary

### Building the Docker Image

To build the Docker image with the exporter, run:
```
$ make container
```

Note: go is not required, as the exporter binary is built in a Docker container. See the [Dockerfile](Dockerfile).

### Building the Binary

To build the binary, run:
```
$ make
```

Note: the binary is built for the OS/arch of your machine. To build binaries for other platforms, see the [Makefile](Makefile).

The binary is built with the name `lightning-prometheus-exporter`.

## Credits

Thank you [contributors](https://github.com/platanus/lightning-prometheus-exporter/graphs/contributors)!

<img src="http://platan.us/gravatar_with_text.png" alt="Platanus" width="250"/>

lightning Prometheus Exporter is maintained by [platanus](http://platan.us).

## License

Lightning Prometheus Exporter is © 2018 platanus, spa. It is free software and may be redistributed under the terms specified in the LICENSE file.


CHANGELOG.md
# Changelog

## Unreleased

## 0.3.0

* Add `process` and `go` metrics.
* Add `channels_balance_satoshis` metric

## 0.2.0

* Fix wallet ballance to use a single metric with labels
  `wallet_balance_satoshis`, with `confirmed` and `unconfirmed` labels
* Change default prefix to `lnd`
* Remove unused flags
* Add `channels` metric, with `pending`, `active` and `inactive` labels
* Add `block_height` metric
* Add `synced_to_chain` metric
* Add channel related metrics
  * `channels_limbo_balance_satoshis`
  * `channels_pending`
  * `channels_waiting_close`

## 0.1.0

* Initial release.
* Support to wallet balance metrics
  * `TotalBallance`
  * `ConfirmedBalance`
  * `UnconfirmedBalance`

collector/lightning.go
package collector

import (
	"log"
	"sync"

	"github.com/platanus/lightning-prometheus-exporter/client"
	"github.com/prometheus/client_golang/prometheus"
)

// LightningCollector collects node metrics. It implements prometheus.Collector interface.
type LightningCollector struct {
	lightningClient *client.LightningClient
	metrics         map[string]*prometheus.Desc
	mutex           sync.Mutex
}

// NewLightningCollector creates an LightningCollector.
func NewLightningCollector(lightningClient *client.LightningClient, namespace string) *LightningCollector {
	return &LightningCollector{
		lightningClient: lightningClient,
		metrics: map[string]*prometheus.Desc{
			"wallet_balance_satoshis":         newGlobalMetric(namespace, "wallet_balance_satoshis", "The wallet balance.", []string{"status"}),
			"peers":                           newGlobalMetric(namespace, "peers", "Number of currently connected peers.", []string{}),
			"channels":                        newGlobalMetric(namespace, "channels", "Number of channels", []string{"status"}),
			"block_height":                    newGlobalMetric(namespace, "block_height", "The node’s current view of the height of the best block", []string{}),
			"synced_to_chain":                 newGlobalMetric(namespace, "synced_to_chain", "The node’s current view of the height of the best block", []string{}),
			"channels_limbo_balance_satoshis": newGlobalMetric(namespace, "channel_limbo_balance_satoshis", "The balance in satoshis encumbered in pending channels", []string{}),
			"channels_pending":                newGlobalMetric(namespace, "channel_pending", "The total pending channels", []string{"status", "forced"}),
			"channels_waiting_close":          newGlobalMetric(namespace, "channel_waiting_close", "Channels waiting for closing tx to confirm", []string{}),
			"channels_balance_satoshis":       newGlobalMetric(namespace, "channels_balance_satoshis", "Sum of all channel funds available", []string{}),
		},
	}
}

// Describe sends the super-set of all possible descriptors of node metrics
// to the provided channel.
func (c *LightningCollector) Describe(ch chan<- *prometheus.Desc) {
	for _, m := range c.metrics {
		ch <- m
	}
}

// Collect fetches metrics from the node and sends them to the provided channel.
func (c *LightningCollector) Collect(ch chan<- prometheus.Metric) {
	c.mutex.Lock() // To protect metrics from concurrent collects
	defer c.mutex.Unlock()

	_, err := c.lightningClient.GetStats()
	if err != nil {
		log.Printf("Error getting stats: %v", err)
		return
	}
	walletStats, _ := c.lightningClient.GetWalletStats()
	nodeStats, _ := c.lightningClient.GetInfoStats()
	pendingChannelsStats, _ := c.lightningClient.GetPendingChannelsStats()
	channelBalanceStats, _ := c.lightningClient.GetChannelsBalanceStats()

	ch <- prometheus.MustNewConstMetric(c.metrics["wallet_balance_satoshis"],
		prometheus.GaugeValue, float64(walletStats.UnconfirmedBalance), "unconfirmed")
	ch <- prometheus.MustNewConstMetric(c.metrics["wallet_balance_satoshis"],
		prometheus.GaugeValue, float64(walletStats.ConfirmedBalance), "confirmed")

	ch <- prometheus.MustNewConstMetric(c.metrics["peers"],
		prometheus.GaugeValue, float64(nodeStats.Peers))
	ch <- prometheus.MustNewConstMetric(c.metrics["channels"],
		prometheus.GaugeValue, float64(nodeStats.ActiveChannels), "active")
	ch <- prometheus.MustNewConstMetric(c.metrics["channels"],
		prometheus.GaugeValue, float64(nodeStats.PendingChannels), "pending")
	ch <- prometheus.MustNewConstMetric(c.metrics["channels"],
		prometheus.GaugeValue, float64(nodeStats.InactiveChannels), "inactive")
	ch <- prometheus.MustNewConstMetric(c.metrics["block_height"],
		prometheus.GaugeValue, float64(nodeStats.BlockHeight))
	ch <- prometheus.MustNewConstMetric(c.metrics["synced_to_chain"],
		prometheus.GaugeValue, float64(nodeStats.SyncedToChain))

	ch <- prometheus.MustNewConstMetric(c.metrics["channels_limbo_balance_satoshis"],
		prometheus.GaugeValue, float64(pendingChannelsStats.TotalLimboBalance))
	ch <- prometheus.MustNewConstMetric(c.metrics["channels_pending"],
		prometheus.GaugeValue, float64(pendingChannelsStats.PendingOpenChannels), "opening", "false")
	ch <- prometheus.MustNewConstMetric(c.metrics["channels_pending"],
		prometheus.GaugeValue, float64(pendingChannelsStats.PendingClosingChannels), "closing", "false")
	ch <- prometheus.MustNewConstMetric(c.metrics["channels_pending"],
		prometheus.GaugeValue, float64(pendingChannelsStats.PendingForceClosingChannels), "closing", "true")
	ch <- prometheus.MustNewConstMetric(c.metrics["channels_waiting_close"],
		prometheus.GaugeValue, float64(pendingChannelsStats.WaitingCloseChannels))

	ch <- prometheus.MustNewConstMetric(c.metrics["channels_balance_satoshis"],
		prometheus.GaugeValue, float64(channelBalanceStats.TotalBalance))
}


exporter.go
package main

import (
	"flag"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"

	"github.com/lightningnetwork/lnd/lncfg"
	"github.com/lightningnetwork/lnd/lnrpc"
	"github.com/lightningnetwork/lnd/macaroons"
	"github.com/platanus/lightning-prometheus-exporter/client"
	"github.com/platanus/lightning-prometheus-exporter/collector"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	macaroon "gopkg.in/macaroon.v2"
)

func getEnv(key, defaultValue string) string {
	value, ok := os.LookupEnv(key)
	if !ok {
		return defaultValue
	}
	return value
}

func getEnvBool(key string, defaultValue bool) bool {
	value, ok := os.LookupEnv(key)
	if !ok {
		return defaultValue
	}
	b, err := strconv.ParseBool(value)
	if err != nil {
		log.Fatalf("Environment Variable value for %s must be a boolean", key)
	}
	return b
}

var (
	// Set during go build
	version   string
	gitCommit string

	// maxMsgRecvSize is the largest message our client will receive. We
	// set this to ~50Mb atm.
	maxMsgRecvSize = grpc.MaxCallRecvMsgSize(1 * 1024 * 1024 * 50)

	// Defaults values
	defaultNamespace     = getEnv("NAMESPACE", "lnd")
	defaultListenAddress = getEnv("LISTEN_ADDRESS", ":9113")
	defaultMetricsPath   = getEnv("TELEMETRY_PATH", "/metrics")
	defaultRPCHost       = getEnv("RPC_HOST", "localhost")
	defaultRPCPort       = getEnv("RPC_PORT", "10009")
	defaultTLSCertPath   = getEnv("TLS_CERT_PATH", "/root/.lnd")
	defaultMacaroonPath  = getEnv("MACAROON_PATH", "")
	defaultGoMetrics, _  = strconv.ParseBool(getEnv("GO_METRICS", "false"))

	// Command-line flags
	namespace = flag.String("namespace", defaultNamespace,
		"The namespace or prefix to use in the exported metrics. The default value can be overwritten by NAMESPACE environment variable.")
	listenAddr = flag.String("web.listen-address", defaultListenAddress,
		"An address to listen on for web interface and telemetry. The default value can be overwritten by LISTEN_ADDRESS environment variable.")
	metricsPath = flag.String("web.telemetry-path", defaultMetricsPath,
		"A path under which to expose metrics. The default value can be overwritten by TELEMETRY_PATH environment variable.")
	rpcHost = flag.String("rpc.host", defaultRPCHost,
		"Lightning node RPC host. The default value can be overwritten by RPC_HOST environment variable.")
	rpcPort = flag.String("rpc.port", defaultRPCPort,
		"Lightning node RPC port. The default value can be overwritten by RPC_PORT environment variable.")
	tlsCertPath = flag.String("lnd.tls-cert-path", defaultTLSCertPath,
		"The path to the tls certificate. The default value can be overwritten by TLS_CERT_PATH environment variable.")
	macaroonPath = flag.String("lnd.macaroon-path", defaultMacaroonPath,
		"The path to the read only macaroon. The default value can be overwritten by MACAROON_PATH environment variable.")
	goMetrics = flag.Bool("go-metrics", defaultGoMetrics,
		"Enable process and go metrics from go client library. The default value can be overwritten by GO_METRICS environmental variable.")
)

func main() {
	flag.Parse()

	log.Printf("Starting Lightning Prometheus Exporter Version=%v GitCommit=%v", version, gitCommit)

	var connCfg = getClientConn()
	rpcclient := lnrpc.NewLightningClient(connCfg)

	client, err := client.NewLightningClient(rpcclient)
	if err != nil {
		log.Fatalf("Could not create Lightning Rpc Client: %v", err)
	}

	// registry
	registry := prometheus.NewRegistry()
	registry.MustRegister(collector.NewLightningCollector(client, *namespace))

	if *goMetrics {
		registry.MustRegister(prometheus.NewGoCollector())
		registry.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
	}

	http.Handle(*metricsPath, promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
			<head><title>Lightning Exporter</title></head>
			<body>
			<h1>Lightning Exporter</h1>
			<p><a href='/metrics'>Metrics</a></p>
			</body>
			</html>`))
	})
	log.Fatal(http.ListenAndServe(*listenAddr, nil))
}

func getClientConn() *grpc.ClientConn {
	// Load the specified TLS certificate and build transport credentials
	// with it.
	creds, err := credentials.NewClientTLSFromFile(*tlsCertPath, "")
	if err != nil {
		log.Fatalf("could not find TLS certificate: %v", err)
	}

	// Create a dial options array.
	opts := []grpc.DialOption{
		grpc.WithTransportCredentials(creds),
	}

	// Load the specified macaroon file.
	macBytes, err := ioutil.ReadFile(*macaroonPath)
	if err != nil {
		log.Fatalf("could not find Macaroon: %v", err)
	}

	mac := &macaroon.Macaroon{}
	if err = mac.UnmarshalBinary(macBytes); err != nil {
		log.Fatalf("unable to decode macaroon: %v", err)
	}

	// Now we append the macaroon credentials to the dial options.
	cred := macaroons.NewMacaroonCredential(mac)
	opts = append(opts, grpc.WithPerRPCCredentials(cred))

	// We need to use a custom dialer so we can also connect to unix sockets
	// and not just TCP addresses.
	genericDialer := lncfg.ClientAddressDialer(*rpcPort)
	opts = append(opts, grpc.WithDialer(genericDialer))
	opts = append(opts, grpc.WithDefaultCallOptions(maxMsgRecvSize))

	conn, err := grpc.Dial(*rpcHost, opts...)
	if err != nil {
		log.Fatalf("unable to connect to RPC server: %v", err)
	}

	return conn
}
